server/src/main/java/org/opensearch/cluster/routing/allocation/decider/RemoteStoreMigrationAllocationDecider.java

```diff
@@ -0,0 +1,170 @@
+        DiscoveryNode targetNode = node.node();
+
+        if (compatibilityMode.equals(CompatibilityMode.STRICT)) {
+            // assuming all nodes are of the same type (all remote or all non-remote)
+            return allocation.decision(
+                Decision.YES,
+                NAME,
+                getDecisionDetails(true, shardRouting, targetNode, " for strict compatibility mode")
+            );
+        }
+
+        if (migrationDirection.equals(Direction.REMOTE_STORE) == false) {
+            // docrep migration direction is currently not supported
+            return allocation.decision(
+                Decision.YES,
+                NAME,
+                getDecisionDetails(true, shardRouting, targetNode, " for non remote_store direction")
+            );
+        }
+
+        // check for remote store backed indices
+        IndexMetadata indexMetadata = allocation.metadata().getIndexSafe(shardRouting.index());
+        if (IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.exists(indexMetadata.getSettings())) {
+            remoteStoreBackedIndex = IndexMetadata.INDEX_REMOTE_STORE_ENABLED_SETTING.get(indexMetadata.getSettings());
+        }
+        if (remoteStoreBackedIndex && targetNode.isRemoteStoreNode() == false) {
+            // allocations and relocations must be to a remote node
+            String reason = String.format(
+                Locale.ROOT,
+                " because a remote store backed index's shard copy can only be %s to a remote node",
+                ((shardRouting.assignedToNode() == false) ? "allocated" : "relocated")
+            );
+            return allocation.decision(Decision.NO, NAME, getDecisionDetails(false, shardRouting, targetNode, reason));
+        }
+
+        if (shardRouting.primary()) {
+            return primaryShardDecision(shardRouting, targetNode, allocation);
+        }
+        return replicaShardDecision(shardRouting, targetNode, allocation);
+    }
+
+    // handle scenarios for allocation of a new shard's primary copy
+    private Decision primaryShardDecision(ShardRouting primaryShardRouting, DiscoveryNode targetNode, RoutingAllocation allocation) {
+        if (targetNode.isRemoteStoreNode() == false) {
+            return allocation.decision(Decision.NO, NAME, getDecisionDetails(false, primaryShardRouting, targetNode, ""));
+        }
+        return allocation.decision(Decision.YES, NAME, getDecisionDetails(true, primaryShardRouting, targetNode, ""));
+    }
+
+    private Decision replicaShardDecision(ShardRouting replicaShardRouting, DiscoveryNode targetNode, RoutingAllocation allocation) {
+        if (targetNode.isRemoteStoreNode()) {
+            ShardRouting primaryShardRouting = allocation.routingNodes().activePrimary(replicaShardRouting.shardId());
+            DiscoveryNode primaryShardNode = allocation.nodes().getNodes().get(primaryShardRouting.currentNodeId());
```

Comment:
Just to understand, what will be the value when the primary is not to be find due to node drop and the cluster is doing an failover?